src/map/clif.cpp

```diff
@@ -1787,39 +1826,25 @@ void clif_hominfo( map_session_data *sd, struct homun_data *hd, int flag ){
 #endif
 	p.flee = status->flee;
 	p.amotion = (flag) ? 0 : status->amotion;
-#if PACKETVER >= 20141016
-	// Homunculus HP bar will screw up if the percentage calculation exceeds signed values
-	// Tested maximum: 21474836(=INT32_MAX/100), any value above will screw up the HP bar
-	if( status->max_hp > ( INT32_MAX / 100 ) ){
+	// Homunculus HP and SP bars will screw up if the percentage calculation exceeds signed values
+	// Tested maximum: 21474836(=INT32_MAX/100), any value above will screw up the bars
+	if( status->max_hp > ( std::is_same<decltype(p.hp), uint16>::value ? INT16_MAX : INT32_MAX / 100 ) ){
 		p.hp = status->hp / ( status->max_hp / 100 );
 		p.maxHp = 100;
 	}else{
-		p.hp = status->hp;
-		p.maxHp = status->max_hp;
+		p.hp = static_cast<decltype(p.hp)>(status->hp);
+		p.maxHp = static_cast<decltype(p.maxHp)>(status->max_hp);
 	}
-#else
-	if( status->max_hp > INT16_MAX ){
-		p.hp = status->hp / ( status->max_hp / 100 );
-		p.maxHp = 100;
-	}else{
-		p.hp = status->hp;
-		p.maxHp = status->max_hp;
-	}
-#endif
-	if( status->max_sp > INT16_MAX ){
+	if( status->max_sp > ( std::is_same<decltype(p.sp), uint16>::value ? INT16_MAX : INT32_MAX / 100 ) ){
```

Comment:
if you trying to merge the bugfix who i applied here ->https://github.com/rathena/rathena/pull/8114
when i'm looking to it now, this problem probably caused because p.maxsp is uint16 instead of int16
ranges:
INT16_MAX = +-32767
UINT16_MAX = +65535

Reply:
But according to Lemongrass the INT16_MAX upper cap was there because of the hp/sp bars, not the packet types, unless I'm missing something

Reply:
![image](https://github.com/user-attachments/assets/859a640c-8f8b-46bc-83e3-337a16c48fbb)
i tested, it works
	p.sp = std::min( static_cast<decltype(p.sp)>( status->sp ), std::numeric_limits<decltype(p.sp)>::max() );
	p.maxSp = std::min( static_cast<decltype(p.maxSp)>( status->max_sp ), std::numeric_limits<decltype(p.maxSp)>::max() );

Reply:
> But according to Lemongrass the INT16_MAX upper cap was there because of the hp/sp bars getting broken, not the packet types, unless I'm missing something

the capping is separated by date, older packets suports max uint16
![image](https://github.com/user-attachments/assets/3001b9a5-a594-4a9b-ab81-1e443cb555fd)

actual packet supports uint32
![image](https://github.com/user-attachments/assets/441884ab-f823-4e8f-941a-4dee60042014)

Reply:
i've tried some over status to check
![image](https://github.com/user-attachments/assets/8d10bd77-27c5-4474-9342-accb7cfd1fa5)

![image](https://github.com/user-attachments/assets/ef8aeefe-6b52-45a6-bf54-cd9b3a317821)
![image](https://github.com/user-attachments/assets/ad5b950e-3c97-49ff-9306-efd28819e34b)

Reply:
Hmm interesting, maybe the bars only break on older clients?

Reply:
the question now is, when did it stop?